### PR TITLE
Update to new Uberenv, Now with Spack Environments!

### PR DESCRIPTION
**Pull request by white238**
This updates us to the newest Spack and more importantly, the newest Uberenv. 

Uberenv was updated (thanks @chapman39 !) to do a lot more modern Spack functionality.  For more info go here:

https://github.com/LLNL/uberenv/pull/96

Brief list:

* Now uses Spack Environments (instead of hot-copying compilers.yaml, packages.yaml, and config.yaml)
* `spack activate` has been removed and we now use Spack Views for our Dev Tools
* `--spack-env-file` is now used to point at a `spack.yaml` instead of pointing at a config directory

### CODE REVIEW

**Review by rhornung67 — APPROVED** [review]

**Comment by rhornung67** [comment]
Chock full o' environments goodness!

**Comment by rhornung67** [comment]
@white238 are you going to merge this PR at some point?

**Review by cyrush — COMMENTED** [review]
Related Q: Are python module vs spack views issues worked out / understood?

**Comment by white238** [comment]
> Related Q: Are python module vs spack views issues worked out / understood?

So you just put this section in your devtools `spack.yaml`:

https://github.com/LLNL/axom/blob/4fcd9ce7b4404e6920f00019c42a4fbf9b8f8896/scripts/spack/devtools_configs/blueos_3_ppc64le_ib_p9/spack.yaml#L17-L21

Views are on by default but honestly I don't know where the view directory is created  by default but this puts it at the base in.

The `build_devtools.py` script then links that directory to the `latest` symlink.

This worked flawlessly for me. It does some Spack magic to symlink executables but mainly it updates the `#!` to point at the Spack built python.

The one weirdness I didn't expect was your python executables (shroud and sphinx in our case) dont end up in their own directories but in python's bin.

File: scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+        all: '{compiler.name}-{compiler.version}/{name}-{version}-{hash}'
+    misc_cache: $spack/../misc_cache
+    test_stage: $spack/../test_stage
+    build_stage::
+    - $spack/../build_stage
+
+  # Regular TPLs do not need views
+  view: false
+
+  compilers::
```
Comment (bmhan12) [inline scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml:20]: Minor: I see `::` here and a few other places. I'm surprised spack parses this fine.

File: src/docs/sphinx/dev_guide/updating_tpls.rst [hunk 2]
```diff
@@ -273,22 +273,22 @@ other Axom developers to use during development, in Axom Gitlab CI testing, etc.
...
              testing. If you wish to test and build with a new compiler or 
-             version on a system, modify the appropriate ``compilers.yaml`` 
+             version on a system, modify the appropriate ``spack.yaml`` 
              file.
 
    To change a version of a non-system TPL, go into the 
    ``axom/scripts/spack/packages`` directory. There you will find a 
    sub-directory for each TPL Axom uses. Modify the contents of the Spack
-   package file ``package.py`` in each package sub-directory as needed. 
+   package file ``spack.py`` in each package sub-directory as needed.
```
Comment (bmhan12) [inline src/docs/sphinx/dev_guide/updating_tpls.rst:291]: `package.py`

File: scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml (hunk 1 above)
Comment (white238) [reply scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml:20]: `::` is Spack yaml magic. It says to clear out all previous default values.

File: src/docs/sphinx/dev_guide/updating_tpls.rst (hunk 2 above)
Comment (white238) [reply src/docs/sphinx/dev_guide/updating_tpls.rst:291]: Good catch that is all sorts of wrong.

**Comment by cyrush** [comment]
Awesome, follow up: How do you get the paths to add them to the host config?

Is that something that is custom in the a Axom scripts, or is it something that the spack package handles?

Or is it OK b/c those paths don't change between the spack build and the view creation?

**Comment by white238** [comment]
> Awesome, follow up: How do you get the paths to add them to the host config?
> 
> Is that something that is custom in the a Axom scripts, or is it something that the spack package handles?
> 
> Or is it OK b/c those paths don't change between the spack build and the view creation?

Axom handles that by splitting the Devtools build and the TPL builds. I then use the `latest` symlink in the TPL spack.yaml.  Here:

https://github.com/LLNL/axom/blob/e534338a186d532ba93c56bb8a79bce18f487175/scripts/spack/configs/toss_3_x86_64_ib/spack.yaml#L301-L342

File: host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake [hunk 3]
```diff
@@ -75,23 +75,23 @@ set(BLT_CMAKE_IMPLICIT_LINK_DIRECTORIES_EXCLUDE "/usr/tce/packages/gcc/gcc-4.9.3
 
 # Root directory for generated TPLs
 
-set(TPL_ROOT "/usr/WS1/axom/libs/blueos_3_ppc64le_ib_p9/2023_02_03_15_43_49/clang-10.0.1.1" CACHE PATH "")
+set(TPL_ROOT "/usr/WS1/axom/libs/blueos_3_ppc64le_ib_p9/2023_04_18_15_47_23/clang-10.0.1.1" CACHE PATH "")
 
-set(CONDUIT_DIR "${TPL_ROOT}/conduit-0.8.6" CACHE PATH "")
+set(CONDUIT_DIR "${TPL_ROOT}/conduit-0.8.6-wizqtmncmicwv3wzfzcjuukwmfbqflc2" CACHE PATH "")
```
Comment (kennyweiss) [inline host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake:80]: > `-wizqtmncmicwv3wzfzcjuukwmfbqflc2`

Out of curiosity, is this suffix due to a change in our setup or the new spack?

File: host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake [hunk 4]
```diff
@@ -101,19 +101,19 @@ set(CAMP_DIR "${TPL_ROOT}/camp-2022.03.2" CACHE PATH "")
 
 # Root directory for generated developer tools
 
-set(DEVTOOLS_ROOT "/collab/usr/gapps/axom/devtools/blueos_3_ppc64le_ib_p9/2022_07_12_11_54_06/gcc-8.3.1" CACHE PATH "")
+set(DEVTOOLS_ROOT "/collab/usr/gapps/axom/devtools/blueos_3_ppc64le_ib_p9/2023_04_18_13_41_48/._view/srxt35kojgk77f2222mk6mgv7z5jyyzz" CACHE PATH "")
```
Comment (kennyweiss) [inline host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake:104]: For the future, should we try to move the output up above the hidden `.view` directory?

**Review by bmhan12 — APPROVED** [review]

File: scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+      flags: {}
+      modules: []
+      operating_system: rhel7
+      paths:
+        cc:  /usr/tce/packages/gcc/gcc-8.3.1/bin/gcc
+        cxx: /usr/tce/packages/gcc/gcc-8.3.1/bin/g++
+        f77: /usr/tce/packages/gcc/gcc-8.3.1/bin/gfortran
+        fc:  /usr/tce/packages/gcc/gcc-8.3.1/bin/gfortran
+      spec: gcc@8.3.1.2
+      target: ppc64le
```
Comment (kennyweiss) [inline scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml:46]: Question for future -- do we still need different entries of the same compiler to support CPU and NVCC builds?

File: scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+      buildable: false
+      externals:
+      - spec: llvm@10.0.0+clang
+        prefix: /usr/tce/packages/clang/clang-10.0.0
+    python:
+      version: [3.10.10]
+      buildable: false
+      externals:
+      - spec: python@3.10.10
+        prefix: /collab/usr/gapps/axom/devtools/blueos_3_ppc64le_ib_p9/latest/python-3.10.10
```
Comment (kennyweiss) [inline scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml:339]: Weird -- looks like we now need to bake in the path to our devtools

File: scripts/spack/configs/toss_4_x86_64_ib/spack.yaml
Comment (kennyweiss) [inline scripts/spack/configs/toss_4_x86_64_ib/spack.yaml:1]: Do we test the toss4 setup yet -- e.g. on a CTS-2 machine ?

**Review by kennyweiss — APPROVED** [review]
Thanks @white238 !

File: scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml (hunk 6 above)
Comment (chapman39) [reply scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml:339]: while latest is pointing to a different path, the prefixes looks practically identical as they did before

https://github.com/LLNL/axom/blob/b27dd885e1a1eddd122ca6f00d2b0d6243ca8832/scripts/spack/configs/blueos_3_ppc64le_ib_p9/packages.yaml#L241-L246

File: host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake (hunk 3 above)
Comment (chapman39) [reply host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake:80]: its a changed due to setup

was this
https://github.com/LLNL/axom/blob/b27dd885e1a1eddd122ca6f00d2b0d6243ca8832/scripts/spack/configs/config.yaml#L21-L22

now is
https://github.com/LLNL/axom/blob/e534338a186d532ba93c56bb8a79bce18f487175/scripts/spack/devtools_configs/blueos_3_ppc64le_ib_p9/spack.yaml#L10-L11

Comment (white238) [reply host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake:80]: I personally am ambivalent to this change, but if we want to allow other projects to trigger our CI and use spack upstreaming to not rebuild the full set of libraries, it is my understanding it's required.

Think Umpire making changes and triggering our CI to see if it still passes Axom tests.

File: host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake (hunk 4 above)
Comment (white238) [reply host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@10.0.1.1.cmake:104]: I think this is a duplicate directory that Spack is creating.. there is also a straight `view` directory right under the timedate stamp directory.

We could investigate later if we can remove that duplication.

File: scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml (hunk 5 above)
Comment (white238) [reply scripts/spack/configs/blueos_3_ppc64le_ib_p9/spack.yaml:46]: I was planning on figuring that out, post this PR. But this PR is huge and I was trying to minimize the changes.

File: scripts/spack/configs/toss_4_x86_64_ib/spack.yaml
Comment (white238) [reply scripts/spack/configs/toss_4_x86_64_ib/spack.yaml:1]: rzvernal passed but i havent tried a toss4 specific machine. I will do that in a follow up PR after quartz comes back up.

**Pull request merged**